### PR TITLE
feat(testing): add hypothesis property-based testing

### LINKED ISSUE #188: feat(testing): add property-based testing with hypothesis
## Summary
Add `hypothesis` for property-based testing — auto-generate hundreds of random inputs to find edge cases in validation logic, config parsing, and Pydantic models.

## Rationale
Traditional tests check specific inputs. Hypothesis generates random inputs matching a described shape, finding edge cases you'd never think of. Especially valuable for shard validators and trust-boundary Pydantic models.

## Pros
- Finds edge cases automatically (unicode, empty strings, None, nested structures)
- Shrinks failing inputs to minimal reproduction case
- Complements existing unit tests
- Excellent for validation functions and parsers

## Cons
- Tests are slower (~1-2s each vs milliseconds for unit tests) — mark with `@pytest.mark.slow`
- Requires learning the strategies API
- Can find technically valid but practically irrelevant edge cases

## Blocking
- Blocked by #179 (pytest-xdist — add to same test infrastructure)

## Verification
- `pytest -m hypothesis` runs property-based tests
- At least one test covers shard validation with random inputs
- Tests generate >50 examples per property

### PR DESCRIPTION
**Pull request by ktinubu**
## Summary

- Adds `hypothesis` to `requirements-app.txt` for property-based testing
- Registers `hypothesis` pytest marker in `pyproject.toml`
- Adds an example property-based test (`tests/test_properties.py`) that fuzzes `OmegaConf.create` with arbitrary dictionaries

## Rationale

Property-based testing complements example-based tests by exploring edge cases we wouldn't think to write manually. This is especially valuable for:

- **Config parsing** — Hydra/OmegaConf must handle or cleanly reject arbitrary inputs
- **Pydantic trust boundaries** — models with `strict=True` should never crash on unexpected data
- **Numeric pipelines** — float edge cases (inf, subnormals, large values) are hard to cover by hand

### Pros
- Finds edge-case bugs that example-based tests miss
- Hypothesis shrinks failing inputs to minimal reproducible cases
- Integrates cleanly with pytest (marker-based, works with xdist)

### Cons
- Slower than unit tests (marked `@pytest.mark.slow` so `make test` skips them)
- Non-deterministic by default (mitigated by Hypothesis's database of failing examples)

## Test plan

- [x] `make format` passes (all pre-commit hooks green)
- [ ] `make test-full` passes (includes the new slow/hypothesis-marked test)
- [ ] Verify `pytest -m hypothesis` runs only the property-based test

Closes #188

### CODE REVIEW

File: tests/test_properties.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    """OmegaConf.create should handle or reject arbitrary dicts without crashing."""
+    from omegaconf import OmegaConf
+
+    try:
+        cfg = OmegaConf.create(random_dict)
+        # If it accepts the dict, it should be convertible back
+        OmegaConf.to_container(cfg)
+    except Exception:  # noqa: S110
+        # Rejection is fine — crashing is not
+        pass
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply tests/test_properties.py]: Fixed in 1b5e7f4. Narrowed to `OmegaConfBaseException` and separated `create` from `to_container` so conversion failures propagate.

One correction: `OmegaConfBaseException` isn't exported at the top level — the import path is `from omegaconf.errors import OmegaConfBaseException`.

File: tests/test_properties.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    st.dictionaries(
+        keys=st.text(min_size=1, max_size=50),
+        values=st.one_of(st.integers(), st.text(), st.none(), st.floats(allow_nan=False)),
+        min_size=0,
+        max_size=20,
+    )
+)
+@settings(max_examples=50)
+def test_arbitrary_dict_does_not_crash_omegaconf(random_dict):
+    """OmegaConf.create should handle or reject arbitrary dicts without crashing."""
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply tests/test_properties.py]: Fixed in 1b5e7f4. Added `random_dict: dict[str, int | str | None | float]` and `-> None` to match the rest of the test suite.

[the remark this answers is not in the document]
Comment (ktinubu) [reply tests/test_properties.py]: Fixed in the follow-up commit — the annotation was added.

That said, pyright didn't catch this because it doesn't enforce missing annotations by default, only type mismatches. We've opened #212 to track tightening this:

- **`src/`**: Enforce `reportMissingParameterType` and `reportMissingReturnType`
- **`tests/`**: Keep current strictness — Hypothesis strategies already declare types via `st.*`, and pytest fixtures are injected by name. Gating here adds friction for no safety benefit.

**Comment by ktinubu** [comment]
## Verification Results

- [x] **Hypothesis is importable and available** *(Level 1 — invoked Python import directly)*
  ```bash
  $ python -c "import hypothesis; print(f'hypothesis {hypothesis.__version__}')"
  hypothesis 6.151.9
  ```

- [x] **`hypothesis` pytest marker is registered** *(Level 1 — queried pytest markers)*
  ```bash
  $ pytest --markers 2>&1 | grep hypothesis
  @pytest.mark.hypothesis: property-based tests (hypothesis)
  @pytest.mark.hypothesis: Tests which use hypothesis.
  ```

- [x] **`make format` passes** *(Level 1 — ran all pre-commit hooks)*
  ```bash
  $ make format
  trim trailing whitespace.................................................Passed
  fix end of files.........................................................Passed
  check docstring is first.................................................Passed
  check yaml...............................................................Passed
  detect private key.......................................................Passed
  check that executables have shebangs.....................................Passed
  don't commit to branch...................................................Passed
  check toml...............................................................Passed
  check for case conflicts.................................................Passed
  check for added large files..............................................Passed
  check json...............................................................Passed
  debug statements (python)................................................Passed
  ruff (legacy alias)......................................................Passed
  ruff format..............................................................Passed
  pyright..................................................................Passed
  Validate GitHub Workflows................................................Passed
  Validate pyproject.toml..................................................Passed
  docformatter.............................................................Passed
  interrogate..............................................................Passed
  prettier.................................................................Passed
  shellcheck...............................................................Passed
  mdformat.................................................................Passed
  codespell................................................................Passed
  ```

- [x] **`pytest -m hypothesis` selects only the property-based test** *(Level 1 — ran pytest with marker filter)*
  ```bash
  $ pytest -m hypothesis -v
  collected 19 items / 18 deselected / 1 selected
  tests/test_properties.py::test_arbitrary_dict_does_not_crash_omegaconf PASSED [100%]
  1 passed, 18 deselected in 0.44s
  ```

- [ ] **`make test-full` passes** *(Level 1 — ran full suite; blocked on pre-existing `test_train_eval` ML training test which has been running 10+ min. The new hypothesis test itself passes — verified via `pytest -m hypothesis` above.)*
